pkg/controller/blockdeviceclaim/blockdeviceclaim_controller.go

```diff
@@ -317,48 +276,64 @@ func (r *ReconcileBlockDeviceClaim) deleteClaimedBlockDevice(
+	if err != nil {
+		return fmt.Errorf("error getting claim reference for BDC:%s, %v", instance.ObjectMeta.Name, err)
 	}
-
-	matchingBlockDevices := v1alpha1.BlockDeviceList{}
-	for _, bd := range bdList.Items {
-		// check whether the block device is unclaimed and active
-		if bd.Status.State != ndm.NDMActive ||
-			bd.Status.ClaimState != v1alpha1.BlockDeviceUnclaimed {
-			continue
-		}
-		// check device type
-		if checkDeviceType && bd.Spec.Details.DeviceType != instance.Spec.DeviceType {
-			continue
-		}
-		matchingBlockDevices.Items = append(matchingBlockDevices.Items, bd)
+	bd.Spec.ClaimRef = claimRef
+	bd.Status.ClaimState = apis.BlockDeviceClaimed
+	err = r.client.Update(context.TODO(), bd)
+	if err != nil {
+		return fmt.Errorf("error while updating BD:%s, %v", bd.ObjectMeta.Name, err)
 	}
-	reqLogger.Info("No of matching devices based on Spec : " + strconv.Itoa(len(matchingBlockDevices.Items)))
-	return matchingBlockDevices
+	reqLogger.Info("Block Device " + bd.Name + " claimed")
+	return nil
 }
 
-// selectBlockDevice selects a single BlockDevice from the list of
-// block devices based on the resource requirements
-func selectBlockDevice(blockDeviceList v1alpha1.BlockDeviceList, resourceLists v1.ResourceList) (selected v1alpha1.BlockDevice, foundMatch bool) {
-	foundMatch = false
-	for _, bd := range blockDeviceList.Items {
-		if matchResourceRequirements(bd, resourceLists) {
-			selected = *bd.DeepCopy()
-			foundMatch = true
-			break
-		}
+// GetBlockDevice get block device resource from etcd
+func (r *ReconcileBlockDeviceClaim) GetBlockDevice(name string) (*apis.BlockDevice, error) {
+	bd := &apis.BlockDevice{}
+	err := r.client.Get(context.TODO(),
+		client.ObjectKey{Namespace: "", Name: name}, bd)
+
+	if err != nil {
+		return nil, err
 	}
-	return
+	return bd, nil
 }
 
-// matchResourceRequirements matches a block device with a ResourceList
-func matchResourceRequirements(bd v1alpha1.BlockDevice, list v1.ResourceList) bool {
-	capacity, _ := getRequestedCapacity(list)
-	return bd.Spec.Capacity.Storage >= uint64(capacity)
+// getListofDevices gets the list of block devices on the node to which BlockDeviceClaim is made
+// TODO:
+//  ListBlockDeviceResource in package cmd/ndm_daemonset/controller has the same functionality.
+//  Need to merge these 2 functions.
+func (r *ReconcileBlockDeviceClaim) getListofDevices(hostName string, ManualSelection bool) (*apis.BlockDeviceList, error) {
+
+	//Initialize a deviceList object.
+	listBlockDevice := &apis.BlockDeviceList{
+		TypeMeta: v1.TypeMeta{
+			Kind:       "BlockDevice",
+			APIVersion: "openebs.io/v1alpha1",
+		},
+	}
+
+	opts := &client.ListOptions{}
+	// Set filter option, in our case we are filtering based on hostname/node
+	// Filter is only applied when its auto selection. In Manual selection, all
+	// blockdevices are listed.
+	// TODO for manual selection, instead of listing all BDs, only get the BD with given name
+	if !ManualSelection {
```

Comment:
Can we change this to specify the hostName filter -- only when hostName is specified. In this way, even in automatic selection it will be possible to get a BD that is available anywhere in the cluster.